api/app/files/api.py

```diff
@@ -45,9 +46,48 @@ def get(self):
             return send_file(temp.name, as_attachment=True, attachment_filename=args['filename'], mimetype='application/pdf')
 
 
+
+class FileAPIImproved(FileUploadMixin, restful.Resource):
+
+
+    @staticmethod
+    def get_by_id(file_id):
+        return db.session.query(File).get(file_id)
+
+    def get(self):
+        # TO DO: Query the File table
+
+        req_parser = reqparse.RequestParser()
+        req_parser.add_argument('file_id', type=int, required=True)
+        args = req_parser.parse_args()
+        LOGGER.debug("FileUpload GET args: {}".format(args))
+
+        bucket = _get_storage_bucket()
+        file = get_by_id(args['file_id'])
+
+        blob = bucket.blob(args['file_id'])
+
+        # TO DO : Rename the file and set the appropriate mime_type
+
+        with tempfile.NamedTemporaryFile() as temp:
+            blob.download_to_filename(temp.name)
+            return send_file(temp.name, as_attachment=True, attachment_filename=file.file_name, mimetype=file.mime_type)
+
+    
+    @staticmethod
+    def save_file_object(file_obj):
+        db.session.add(file_obj)
+        db.session.commit()
+        last_save_file = db.session.query(File).last()
+        return last_save_file.file_id

```

Comment:
As described elsewhere, move this into a repository file and class.

Also the way you're retrieving the file_id is risky. What if two people upload files at almost exactly the same time? It's possible the first person will get back the second persons `file_id` which is incorrect.

Instead I think you might be able to just say return file.file_id. SQL Alchemy should hydrate the `file_id` automatically for you (but I think just test and make sure). Also rename `file_obj` as `file` and `save_file_obj` as `save`. We know it's an object :)

Reply:
Hey @amritpurshotam. Finish with new updates. But it seems there is 1 issue. Would you please tell me where it comes from ? Hope it's not a git issue again ? :) !